### PR TITLE
Add clipper deployment support

### PR DESCRIPTION
**Pull request by yubozhao**
(Thanks for sending a pull request! Please make sure to read the contribution guidelines, then fill out the blanks below.)

What changes were proposed in this pull request?
------------------------------------------------

Does this close any currently open issues?
------------------------------------------

How was this patch tested?
--------------------------

### CODE REVIEW

File: bentoml/archive/py_module_utils.py [hunk 1]
```diff
@@ -12,7 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-
```
Comment (parano) [inline bentoml/archive/py_module_utils.py]: ah sorry about this line added earlier, should have run the auto format script, could you make a separate PR for this?

File: .gitignore [hunk 2]
```diff
@@ -109,3 +109,4 @@ package.json
 node_modules/
 yarn-error.log
 yarn.lock
+.DS_Store
```
Comment (parano) [inline .gitignore]: nit:
```
# MacOS
.DS_Store
```

File: bentoml/archive/templates.py [hunk 3]
```diff
@@ -150,3 +149,57 @@ def load():
 
 __all__ = ['__version__', '{service_name}', 'load']
 """
+
+BENTO_SERVICE_DOCKERFILE_CLIPPER_TEMPLATE = """\
+FROM clipper/python36-closure-container:0.3
```
Comment (parano) [inline bentoml/archive/templates.py]: add a note on why we use python36 image, even for BentoArchive built in python27

File: bentoml/archive/templates.py [hunk 4]
```diff
@@ -150,3 +149,57 @@ def load():
...
+    /bin/bash ~/miniconda.sh -b -p /opt/conda && \
+    rm ~/miniconda.sh && \
+    /opt/conda/bin/conda clean -tipsy && \
+    ln -s /opt/conda/etc/profile.d/conda.sh /etc/profile.d/conda.sh && \
+    echo ". /opt/conda/etc/profile.d/conda.sh" >> ~/.bashrc && \
+    echo "conda activate base" >> ~/.bashrc
+
+ENV TINI_VERSION v0.16.1
+ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /usr/bin/tini
+RUN chmod +x /usr/bin/tini
```
Comment (parano) [inline bentoml/archive/templates.py]: do we need tini in this case?

File: bentoml/archive/templates.py [hunk 5]
```diff
@@ -150,3 +149,57 @@ def load():
...
+##### End installation for miniconda3
+
+RUN set -x \
```
Comment (parano) [inline bentoml/archive/templates.py]: move this up to the "apt-get" section above, let's not run `apt-get update`  multiple times in one docker build

File: bentoml/archive/templates.py [hunk 6]
```diff
@@ -150,3 +149,57 @@ def load():
...
+      && conda install pip numpy scipy \
+      && pip install six bentoml
+
+# copy over model files
+COPY ./bento /container/bento
+COPY ./clipper_entry.py /container/clipper_entry.py
+WORKDIR /container
+
+# update conda base env
+RUN conda env update -n base -f /container/environment.yml
```
Comment (parano) [inline bentoml/archive/templates.py]: shouldn't this be "/container/bento/environment.yml"?

Comment (parano) [reply bentoml/archive/templates.py]: same for the requirement.txt and setup.sh file below?

File: bentoml/deployment/clipper/__init__.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+        entry_py_content = DEFAULT_CLIPPER_ENTRY.format(api_name=self.api.name,
+                                                        input_type=self.input_type)
+        with open(os.path.join(snapshot_path, 'clipper_entry.py'), 'w') as f:
+            f.write(entry_py_content)
+        model_path = os.path.join(snapshot_path, 'bento')
+        shutil.copytree(self.archive_path, model_path)
+
+        image = build_docker_image(snapshot_path)
+        self.clipper_conn.register_application(name=self.application_name,
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: I would suggest leaving register_application and link_model_to_app to the user, similar to the "deploy_xxx" commands provided in clipper, it should only deploy the model, without affecting the application

File: bentoml/deployment/clipper/__init__.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        self.clipper_conn.deploy_model(name=self.application_name,
+                                       version=self.bento_service.version,
+                                       input_type=self.input_type, image=image)
+
+        self.clipper_conn.link_model_to_app(self.application_name, self.application_name)
+        self.is_deployed = True
+        return
+
+    def check_status(self):
+        if self.is_deployed:
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: discussed offline, remove check_status and delete here, will recommend user follow clipper convention on checking the model status and stopping running model instance

File: bentoml/archive/archiver.py [hunk 9]
```diff
@@ -148,6 +148,8 @@ def _save(bento_service, dst, version=None):
         f.write(BENTO_SERVICE_DOCKERFILE_CPU_TEMPLATE)
     with open(os.path.join(path, 'Dockerfile-sagemaker'), 'w') as f:
         f.write(BENTO_SERVICE_DOCKERFILE_SAGEMAKER_TEMPLATE)
+    with open(os.path.join(path, 'Dockerfile-clipper'), 'w') as f:
+        f.write(BENTO_SERVICE_DOCKERFILE_CLIPPER_TEMPLATE)
```
Comment (parano) [inline bentoml/archive/archiver.py]: discussed offline, we should handle this similar to DEFAULT_CLIPPER_ENTRY, generate it when calling deploy

Comment (parano) [reply bentoml/archive/archiver.py]: same for BENTO_SERVICE_DOCKERFILE_SAGEMAKER_TEMPLATE above, but this can be in separate PR

File: bentoml/deployment/clipper/__init__.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    for line in docker_api.build(path=snapshot_path, dockerfile='Dockerfile-clipper', tag=tag):
+        process_docker_api_line(line)
+    return tag
+
+
+class ClipperDeployment(Deployment):
+    """Clipper deployment for BentoML bundle
+    """
+
+    def __init__(self, clipper_conn, archive_path, api_name, slo_micros=100000):
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: remove slo_micros since we will not handle register application

File: bentoml/deployment/clipper/__init__.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        image = build_docker_image(snapshot_path)
+
+        self.clipper_conn.deploy_model(name=self.application_name,
+                                       version=self.bento_service.version,
+                                       input_type=self.input_type, image=image)
+
+        return self.application_name
+
+
+def deploy_bentoml(clipper_conn, archive_path, api_name):
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: also expose the clipper "labels" here? http://clipper.ai/api/api.html#definition-Model

File: bentoml/archive/templates.py (hunk 5 above)
Comment (yubozhao) [reply bentoml/archive/templates.py]: done

File: bentoml/archive/templates.py (hunk 4 above)
Comment (yubozhao) [reply bentoml/archive/templates.py]: na, removed it.

done

File: bentoml/archive/templates.py (hunk 6 above)
Comment (yubozhao) [reply bentoml/archive/templates.py]: yeah, those weren't done when i made the draft PR.

done now

File: bentoml/deployment/clipper/__init__.py (hunk 7 above)
Comment (yubozhao) [reply bentoml/deployment/clipper/__init__.py]: done

File: bentoml/deployment/clipper/__init__.py (hunk 8 above)
Comment (yubozhao) [reply bentoml/deployment/clipper/__init__.py]: done

File: bentoml/archive/archiver.py (hunk 9 above)
Comment (yubozhao) [reply bentoml/archive/archiver.py]: done

File: bentoml/deployment/clipper/__init__.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+from bentoml.handlers import ImageHandler
+from bentoml.deployment.utils import generate_bentoml_deployment_snapshot_path, \
+    process_docker_api_line
+from bentoml.deployment.clipper.templates import DEFAULT_CLIPPER_ENTRY, DOCKERFILE_CLIPPER
+from bentoml.utils.exceptions import BentoMLException
+
+
+def build_docker_image(snapshot_path):
+    docker_api = docker.APIClient()
+    tag = "something"
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: maybe use BentoArchive version for tag?

File: bentoml/deployment/clipper/__init__.py (hunk 10 above)
Comment (yubozhao) [reply bentoml/deployment/clipper/__init__.py]: done

File: bentoml/deployment/clipper/__init__.py (hunk 12 above)
Comment (yubozhao) [reply bentoml/deployment/clipper/__init__.py]: done

File: bentoml/archive/py_module_utils.py (hunk 1 above)
Comment (yubozhao) [reply bentoml/archive/py_module_utils.py]: yep, will do that

File: bentoml/deployment/clipper/__init__.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+    if api_name:
+        api = next(item for item in apis if item.name == api_name)
+    elif len(apis) == 1:
+        api = apis[0]
+    else:
+        raise BentoMLException(
+            'Please specify api-name, when more than one API is present in the archive')
+    application_name = bento_service.name + '-' + api.name
+    input_type = 'strings'
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: Why do we still need 'handle_clipper_numbers'? looks like its just strings and bytes we are supporting?

File: .gitignore (hunk 2 above)
Comment (yubozhao) [reply .gitignore]: done

File: bentoml/deployment/clipper/__init__.py (hunk 11 above)
Comment (yubozhao) [reply bentoml/deployment/clipper/__init__.py]: done

File: bentoml/handlers/dataframe_handler.py [hunk 14]
```diff
@@ -195,3 +196,72 @@ def handle_aws_lambda_event(self, event, func):
             result, event["headers"].get("output", "json"), output_orient
         )
         return {"statusCode": 200, "body": result}
+
+    def handle_clipper_strings(self, inputs, func):
+        def transform_and_predict(input_string):
+            # Assuming input string is JSON format
```
Comment (yubozhao) [inline bentoml/handlers/dataframe_handler.py]: Maybe give user a choice??

File: bentoml/deployment/clipper/__init__.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+def deploy_bentoml(
+    clipper_conn, archive_path, api_name, input_type='strings', labels=["bentoml"]
+):
+    """Deploy bentoml bundle to clipper cluster
+
+    Args:
+        clipper_conn(clipper_admin.ClipperConnection): Clipper conntection instance
+        archive_path(str): Path to the bentoml service archive.
+        api_name(str): name of the api that will be running in the clipper cluster
+        input_type(str): Input type that clipper accept
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: add some more notes here - especially about the default value and behavior around ImageHnalder check below

File: bentoml/deployment/clipper/__init__.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+    if api_name:
+        api = next(item for item in apis if item.name == api_name)
+    elif len(apis) == 1:
+        api = apis[0]
+    else:
+        raise BentoMLException(
+            "Please specify api-name, when more than one API is present in the archive"
+        )
+    application_name = generate_clipper_compatiable_string(
```
Comment (parano) [inline bentoml/deployment/clipper/__init__.py]: give the user the option to provide an application name as the function parameter?

File: bentoml/handlers/dataframe_handler.py [hunk 17]
```diff
@@ -72,15 +73,15 @@ def __init__(
...
-        if item.startswith('int'):
-            return 'integer'
-        if item.startswith('float') or item.startswith('double'):
-            return 'number'
-        if item.startswith('str') or item.startswith('date'):
-            return 'string'
-        if item.startswith('bool'):
-            return 'boolean'
-        return 'object'
+        if item.startswith("int"):
```
Comment (parano) [inline bentoml/handlers/dataframe_handler.py]: is this formatted by black? I thought it should be single quote?

Comment (yubozhao) [reply bentoml/handlers/dataframe_handler.py]: Yes, those formatted by black.  I will re-run it to make sure.

File: bentoml/deployment/clipper/__init__.py (hunk 16 above)
Comment (yubozhao) [reply bentoml/deployment/clipper/__init__.py]: done

File: bentoml/deployment/clipper/__init__.py (hunk 15 above)
Comment (yubozhao) [reply bentoml/deployment/clipper/__init__.py]: done

File: bentoml/handlers/dataframe_handler.py (hunk 17 above)
Comment (yubozhao) [reply bentoml/handlers/dataframe_handler.py]: changed back.  format script skip string normalization

File: bentoml/service.py [hunk 18]
```diff
@@ -105,6 +105,21 @@ def handle_cli(self, args):
     def handle_aws_lambda_event(self, event):
         return self.handler.handle_aws_lambda_event(event, self.func)
 
+    def handle_clipper_strings(self, inputs):
+        return self.handler.handle_clipper_strings(inputs, self.func)
+
+    def handle_clipper_bytes(self, inputs):
+        return self.handler.handle_clipper_strings(inputs, self.func)
```
Comment (parano) [inline bentoml/service.py]: typo

File: bentoml/handlers/image_handler.py [hunk 19]
```diff
@@ -166,3 +166,39 @@ def handle_aws_lambda_event(self, event, func):
...
+
+        def transform_and_predict(input_bytes):
+            data = cv2.imdecode(input_bytes, cv2.IMREAD_COLOR)
+            return func(data)
+
+        return list(map(transform_and_predict, inputs))
+
+    def handle_clipper_strings(self, inputs, func):
+        raise RuntimeError(
+            "Image handler does not support 'strings' input_type \
```
Comment (parano) [inline bentoml/handlers/image_handler.py]: typo? *ImageHandler

File: bentoml/handlers/dataframe_handler.py [hunk 20]
```diff
@@ -195,3 +196,72 @@ def handle_aws_lambda_event(self, event, func):
...
+                    input_string, orient=self.orient, typ=self.typ, dtype=False
+                )
+            except ValueError as e:
+                raise ValueError(
+                    "Unexpected input format, BentoML DataframeHandler expects json "
+                    "string as input: {}".format(e)
+                )
+            return func(df)
+
+        result = map(transform_and_predict, inputs)
```
Comment (parano) [inline bentoml/handlers/dataframe_handler.py]: nit: just do `return list(map(...))`

File: bentoml/deployment/utils.py [hunk 21]
```diff
@@ -32,3 +34,23 @@ def generate_bentoml_deployment_snapshot_path(service_name, service_version, pla
...
+                line_payload = json.loads(line)
+            except ValueError as e:
+                print("Could not decipher payload from Docker API: " + e)
+            if line_payload:
+                if "errorDetail" in line_payload:
+                    error = line_payload["errorDetail"]
+                    sys.stderr.write(error["message"])
+                    raise RuntimeError("Error on build - code %s" % error["code"])
+                elif "stream" in line_payload:
+                    # TODO: move this to logger.info
```
Comment (parano) [inline bentoml/deployment/utils.py]: try just use logger.info here? and similarly for the logging.error above?

File: bentoml/deployment/clipper/templates.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+    /opt/conda/bin/conda clean -tipsy && \
+    ln -s /opt/conda/etc/profile.d/conda.sh /etc/profile.d/conda.sh && \
+    echo ". /opt/conda/etc/profile.d/conda.sh" >> ~/.bashrc && \
+    echo "conda activate base" >> ~/.bashrc
+
+
+# update conda and setup environment and pre-install common ML libraries to speed up docker build
+RUN conda update conda -y \
+      && conda install pip numpy scipy \
+      && pip install six bentoml
```
Comment (parano) [inline bentoml/deployment/clipper/templates.py]: this is ok for now but we need to pin this to the version that user is using for saving the given BentoArchive

File: bentoml/deployment/utils.py (hunk 21 above)
Comment (yubozhao) [reply bentoml/deployment/utils.py]: done
